File: 0375-9-2874-maximum-value-of-an-ordered-triplet-ii/0375-9-2874-maximum-value-of-an-ordered-triplet-ii.java
class Solution {
    public long maximumTripletValue(int[] nums) {
        long min = 0;
        long min2 = 0;
        long mul = 0;

        for(int i=0;i<nums.length;i++){
            mul = Math.max(mul, min * nums[i]);
            min = Math.max(min, min2 - nums[i]);
            min2 = Math.max(min2, nums[i]);
        }

        return mul;
    }
}

File: 0375-9-2874-maximum-value-of-an-ordered-triplet-ii/0375-9-2874-maximum-value-of-an-ordered-triplet-ii.py
class Solution:
    def maximumTripletValue(self, nums: List[int]) -> int:
        res = 0
        imax = 0
        dmax = 0
        for i in range(len(nums)):
            res = max(res, dmax * nums[i])
            dmax = max(dmax, imax - nums[i])
            imax = max(imax, nums[i])
        return res